src/libsyntax/parse/unescape.rs

```diff
@@ -0,0 +1,512 @@
+    let mut chars = literal_text.chars();
+    let res = (|| {
+        let first_char = chars.next().ok_or(EscapeError::ZeroChars)?;
+        let res = scan_escape(first_char, &mut chars, Mode::Char)?;
+        if chars.next().is_some() {
+            return Err(EscapeError::MoreThanOneChar);
+        }
+        Ok(res)
+    })();
+    res.map_err(|err| (literal_text.len() - chars.as_str().len(), err))
+}
+
+/// Takes a contents of a string literal (without quotes) and produces a
+/// sequence of escaped characters or errors.
+pub(crate) fn unescape_str<F>(literal_text: &str, callback: &mut F)
+where
+    F: FnMut(Range<usize>, Result<char, EscapeError>),
+{
+    unescape_str_or_byte_str(literal_text, Mode::Str, callback)
+}
+
+pub(crate) fn unescape_byte(literal_text: &str) -> Result<u8, (usize, EscapeError)> {
+    let mut chars = literal_text.chars();
+    let res = (|| {
+        let first_char = chars.next().ok_or(EscapeError::ZeroChars)?;
+        let res = scan_escape(first_char, &mut chars, Mode::Byte)?;
+        if chars.next().is_some() {
+            return Err(EscapeError::MoreThanOneChar);
+        }
+        Ok(byte_from_char(res))
+    })();
+    res.map_err(|err| (literal_text.len() - chars.as_str().len(), err))
+}
+
+/// Takes a contents of a string literal (without quotes) and produces a
+/// sequence of escaped characters or errors.
+pub(crate) fn unescape_byte_str<F>(literal_text: &str, callback: &mut F)
+where
+    F: FnMut(Range<usize>, Result<u8, EscapeError>),
+{
+    unescape_str_or_byte_str(literal_text, Mode::ByteStr, &mut |range, char| {
+        callback(range, char.map(byte_from_char))
+    })
+}
+
+fn byte_from_char(c: char) -> u8 {
+    let res = c as u32;
+    assert!(res <= u8::max_value() as u32, "guaranteed because of Mode::Byte");
+    res as u8
+}
+
+#[derive(Clone, Copy)]
+enum Mode {
+    Char,
+    Str,
+    Byte,
+    #[allow(unused)]
+    ByteStr,
+}
+
+impl Mode {
+    fn in_single_quotes(self) -> bool {
+        match self {
+            Mode::Char | Mode::Byte => true,
+            Mode::Str | Mode::ByteStr => false,
+        }
+    }
+
+    fn in_double_quotes(self) -> bool {
+        !self.in_single_quotes()
+    }
+
+    fn allow_unicode(self) -> bool {
+        match self {
+            Mode::Byte | Mode::ByteStr => false,
+            Mode::Char | Mode::Str => true,
+        }
+    }
+}
+
+fn scan_escape(first_char: char, chars: &mut Chars<'_>, mode: Mode) -> Result<char, EscapeError> {
+    if first_char != '\\' {
+        return match first_char {
+            '\t' | '\n' => Err(EscapeError::EscapeOnlyChar),
+            '\r' => Err(if chars.clone().next() == Some('\n') {
+                EscapeError::EscapeOnlyChar
+            } else {
+                EscapeError::BareCarriageReturn
+            }),
+            '\'' if mode.in_single_quotes() => Err(EscapeError::EscapeOnlyChar),
+            '"' if mode.in_double_quotes() => Err(EscapeError::EscapeOnlyChar),
```

Comment:
Seems like these two lines are unreachable (if `scan_escape` is called from the real lexer, rather than unit tests).

Reply:
One would think so, but lexer knows to tokenize `'''` as a `char`, so as to provide a targeted error message.

Reply:
Ah, ok.